apps/arena/lib/arena/game/skill.ex

```diff
@@ -217,15 +242,94 @@ defmodule Arena.Game.Skill do
 
       Process.send_after(self(), {:remove_effect, player.id, last_id}, effect.duration_ms)
 
-      effects =
-        get_in(game_state, [:players, player.id, :aditional_info, :effects])
-        |> Map.put(last_id, effect)
-
-      put_in(game_state, [:players, player.id, :aditional_info, :effects], effects)
+      put_in(
+        game_state,
+        [:players, player.id, :aditional_info, :effects, last_id],
+        Map.put(effect, :id, last_id)
+      )
       |> put_in([:last_id], last_id)
     end)
   end
 
+  def apply_effect_mechanic(players, game_state) do
+    updated_players =
+      Map.filter(players, fn {_, player} -> Player.alive?(player) end)
+      |> Enum.map(fn {_player_id, player} ->
+        player =
+          Enum.reduce(player.aditional_info.effects, player, fn {_effect_id, effect}, player ->
+            apply_effect_mechanic(player, effect, game_state)
+          end)
+
+        {player.id, player}
+      end)
+      |> Map.new()
+
+    Map.merge(players, updated_players)
+  end
+
+  def apply_effect_mechanic(player, effect, game_state) do
+    now = DateTime.utc_now() |> DateTime.to_unix(:millisecond)
+
+    mechanics_to_apply =
+      effect.effect_mechanics
+      |> Map.filter(fn {_name, mechanic} ->
+        now - Map.get(mechanic, :last_application_time, 0) >= mechanic.effect_delay_ms
+      end)
+      |> Enum.map(fn {name, mechanic} ->
+        {name, Map.put(mechanic, :last_application_time, now)}
+      end)
+      |> Map.new()
+
+    Enum.reduce(mechanics_to_apply, player, fn mechanic, player ->
+      do_effect_mechanics(game_state, player, effect, mechanic)
+    end)
+    |> update_in([:aditional_info, :effects, effect.id, :effect_mechanics], fn mechanics ->
+      Map.merge(mechanics, mechanics_to_apply)
+    end)
+  end
```

Comment:
So in this function I'm counting 3 iterations over the same list, although performance wise certainly not a problem I think we can do better cause the code becomes confusing

In essence you have 2 behaviours, either enough time has passed and you apply the effect mechanic or you don't, so how about this

```elixir 
  def apply_effect_mechanic(player, effect, game_state) do
    now = System.monotonic_time(:millisecond)
    
    Enum.reduce(effect.effect_mechanics, player, fn {_, mechanic}, player_acc ->
      last_application_at = Map.get(mechanic, :last_application_time, 0)
      
      if now - last_application_at >= mechanic.effect_delay_ms do
        do_effect_mechanics(game_state, player_acc, effect, mechanic)
        |> Map.put_in([:aditional_info, :effects, effect.id, :effect_mechanics, name, :last_application_time], now)
      else
        player_acc
      end
    end)
  end
```

Reply:
I think we should be a little concerned about in performance in this type of methods since this is executed on every tick, great comment!